Classes/MBURLConnectionOperation.m

```diff
@@ -153,142 +166,128 @@ - (void)handleResponse
+    if (!self.cancelled && !self.finished)
     {
-        if ([self shouldCancel])
+        if (self.shouldCancel)
         {
             [self cancelFromRunLoop];
         }
         else
         {
-            [self setResponseData:[NSData dataWithData:[self incrementalResponseData]]];
-            [self setIncrementalResponseData:nil];
-            [self handleResponse];
-            [self finish];
+            if ([self.delegate respondsToSelector:@selector(connectionOperation:didSendBodyData:totalBytesWritten:totalBytesExpectedToWrite:)])
+            {
+                [self.delegate connectionOperation:self
+                               didSendBodyData:(NSUInteger)bytesSent
+                             totalBytesWritten:(NSUInteger)totalBytesSent
+                     totalBytesExpectedToWrite:totalBytesExpectedToSend];
+            }
         }
     }
 }
 
-- (NSCachedURLResponse *)connection:(NSURLConnection *)connection willCacheResponse:(NSCachedURLResponse *)cachedResponse
+- (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didCompleteWithError:(nullable NSError *)error
 {
-    return ([self isCancelled] || [self shouldCancel]) ? nil : cachedResponse;
-}
-
-- (void)connection:(NSURLConnection *)connection
-   didSendBodyData:(NSInteger)bytesWritten
- totalBytesWritten:(NSInteger)totalBytesWritten
-totalBytesExpectedToWrite:(NSInteger)totalBytesExpectedToWrite
-{
-    if (![self isCancelled] && ![self isFinished])
+    if (!self.cancelled && !self.finished)
     {
-        if ([self shouldCancel])
+        if (self.shouldCancel)
         {
             [self cancelFromRunLoop];
         }
         else
         {
-            if ([_delegate respondsToSelector:@selector(connectionOperation:didSendBodyData:totalBytesWritten:totalBytesExpectedToWrite:)])
+            if (error == nil)  // If no error, finished sucesfully
             {
-                [_delegate connectionOperation:self
-                               didSendBodyData:(NSUInteger)bytesWritten
-                             totalBytesWritten:(NSUInteger)totalBytesWritten
-                     totalBytesExpectedToWrite:totalBytesExpectedToWrite];
+                self.responseData = [NSData dataWithData:self.incrementalResponseData];
+                self.incrementalResponseData = nil;
             }
+            else
+            {
+                self.error = error;
+                MBRequestLog(@"Request Error: %@", [self error]);
+            }
+            [self handleResponse];
+            [self finish];
         }
     }
 }
 
-- (void)connection:(NSURLConnection *)connection willSendRequestForAuthenticationChallenge:(NSURLAuthenticationChallenge *)challenge
+- (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didReceiveChallenge:(NSURLAuthenticationChallenge *)challenge completionHandler:(void (^)(NSURLSessionAuthChallengeDisposition, NSURLCredential * _Nullable))completionHandler
```

Comment:
@logancautrell   Well, looks like for "non-session level challenges" (I'm assuming task level ones), `URLSession:task:didReceiveChallenge:completionHandler:` gets called. So I believe this is needed.

<img width="852" alt="screen shot 2016-05-24 at 7 42 11 pm" src="https://cloud.githubusercontent.com/assets/8488627/15507304/b7c15f58-21e8-11e6-8796-65cd07761e4e.png">